project-koku/koku
# Fixing order_by bug and adding order_by account alias support

AWS aliases are to be used in place of `account_usage_id` For the purposes of the API this ordering has to be done manually since it's not part of the AWS report schema.  The reason why AWS aliases are not part of the report schema is because the alias is not part of the report and we shouldn't be adding foreign values into it.  Another practical reason is that the alias can be changed out of band with cost usage report updates.  Since a customer can change their account alias we would want the koku UI to reflect the most current known alias even if a new cost and usage report has not been downloaded.

For this implementation I'm simply doing the ordering manually the only record in the database that we are maintaining is a mapping for account_id to account_alias.  If an account_alias is not known the account_id is used for the purposes of ordering.

**Improvement Areas**
* When group_by account is not in the query and order_by account_alias is, we should probably throw out the order_by query all together.  In the interest of time and priority, I didn't bother implementing that in this change.

**Testing**
1. Add data to koku database where some contain AWS aliases and others that do not.
2. Hit a reports/costs API where group_by account is included and order_by account_alias is desc.  Verify that the values are returned in account_alias in decending order.
3. Hit a reports/costs API where group_by account is included and order_by account_alias is asc.  Verify that the values are returned in account_alias in ascending order.
4. Remove the order_by account_alias and verify that the default order_by is respected.  Default is 'total': 'desc'
5. Issue an order_by account alias without grouping by accounts and verify that the order_by account_alias is ignored

**Test Results**
[koku_order_by_alias.txt](https://github.com/project-koku/koku/files/2415076/koku_order_by_alias.txt)


Closes #375 
Closes #326
